lib/honeybadger/config_sync_service.rb

```diff
@@ -0,0 +1,74 @@
+require 'honeybadger/check_in'
+module Honeybadger
+  class ConfigSyncService
+    def initialize(config)
+      @config = config
+      @check_ins_cache = nil
+    end
+
+    def sync_checkins
+      checkin_configs = @config.get(:checkins) || []
+      return [] if checkin_configs.empty?
+
+      checkins = checkin_configs.map do |cfg|
+        check_in = CheckIn.from_config(cfg)
+        check_in.validate!
+        check_in
+      end
+      created_or_updated = sync_existing_checkins(checkins)
+      removed = sync_removed_checkins(checkins)
+
+      return (created_or_updated + removed).uniq
+    end
+
+    private
+
+    def get_checkin_by_name(project_id, name)
+      @check_ins_cache ||= @config.backend.get_check_ins(project_id)
+      @check_ins_cache.find {|c| c.name == name }
+    end
+
+    def sync_existing_checkins(checkins = [])
+      return [] if checkins.empty?
+      result = []
+      checkins.each do |check_in|
+        remote_checkin = if check_in.id
+          @config.backend.get_check_in(check_in.project_id, check_in.id)
+        else
+          get_checkin_by_name(check_in.project_id, check_in.name)
+        end
+        if remote_checkin
+          unless remote_checkin == check_in
+            result << @config.backend.update_check_in(check_in.project_id, remote_checkin.id, check_in)
+          end
+        else
+          result << @config.backend.create_check_in(check_in.project_id, check_in)
+        end
+      end
+      result
+    end
+
+    def sync_removed_checkins(checkins)
+      return [] if checkins.nil? || checkins.empty?
+      result = []
+      project_ids = checkins.map{|ch| ch.project_id }.uniq
+      project_ids.each do |prj_id|
+        @check_ins_cache ||= @config.backend.get_check_ins(prj_id)
+
+        local_project_checkins = checkins.select {|c| c.project_id == prj_id }
+        to_remove = @check_ins_cache.reject do |pc|
+          local_project_checkins.find{|c| c.id == pc.id || c.name == pc.name }
+        end
+        to_remove.each do |ch|
+          if ch.id.nil?
```

Comment:
This shouldn't happen:
- If a check-in is to be removed, it means that it is available remotely (i.e. it came from the HB API) but does not exist in the local config array
- Since it came from the HB API, it must have an id